web/src/main/java/org/wso2/testgrid/web/api/ProductService.java

```diff
@@ -124,4 +139,60 @@ public Response getAllProductStatuses() {
         }
         return Response.status(Response.Status.OK).entity(list).build();
     }
+
+    /**
+     * This method is able to get the latest report of a given product from the remote storage and return.
+     * <p>
+     * <p> The report is returned as a html file<p/>
+     *
+     * @return latest report of querying product
+     */
+    @GET @Path("/reports") @Produces("application/html") public Response getProductReport(
+            @QueryParam("product-name") String productName,
+            @DefaultValue("false") @QueryParam("show-success") Boolean showSuccess,
+            @DefaultValue("SCENARIO") @QueryParam("group-by") String groupBy) {
+        try {
+            ProductUOW productUOW = new ProductUOW();
+            Optional<Product> productInstance = productUOW.getProduct(productName);
+            Product product;
+            if (productInstance.isPresent()) {
+                product = productInstance.get();
+                AxisColumn uniqueAxisColumn = AxisColumn.valueOf(groupBy.toUpperCase(Locale.ENGLISH));
+                String fileName = StringUtil
+                        .concatStrings(product.getName(), "-", uniqueAxisColumn, Constants.HTML_EXTENSION);
+                String bucketKey = Paths.get(Constants.AWS_BUCKET_ARTIFACT_DIR, productName, fileName).toString();
+                ArtifactReadable artifactDownloadable = new AWSArtifactReader(ConfigurationContext.
+                        getProperty(ConfigurationContext.ConfigurationProperties.AWS_REGION_NAME),
+                        Constants.AWS_BUCKET_NAME);
+                InputStream inputStream = artifactDownloadable.getArtifactStream(bucketKey);
+                Response.ResponseBuilder response = Response.ok(IOUtils.toByteArray(inputStream));
```

Comment:
When you convert the inputstream into a byte array, the entire report will be built in-memory. It seems ok since the report size is small, but can we not directly provide the inputstream to the jax-rs provider? In that case, we can support better download speed by streaming the content.

wdyt?

Reply:
+1 fixed with 751dd40f5a5e69ca20eef7775268e501116da992